sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala

```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
+                      inputTimesMicrosCv =>
+                        inputTimesMicrosCv.castTo(DType.TIMESTAMP_MICROSECONDS)
+                    }
+                  }
+                } else {
+                  withResource(inputWithoutNanAndInfinity.mul(microsPerSec, DType.INT64)) {
+                    inputTimesMicrosCv =>
+                      inputTimesMicrosCv.castTo(DType.TIMESTAMP_MICROSECONDS)
+                  }
+                }
+            }
+          }
+        }
+      }
+    case dt: DecimalType =>
+      (input: GpuColumnVector) => {
+        println("DecimalType " + dt.precision + " " + dt.scale)
+        val decimal128Type = DType.create(DType.DTypeEnum.DECIMAL128, -6)
+        val decimal128TypeAllScale = DType.create(DType.DTypeEnum.DECIMAL128, -dt.scale)
+        // Match the behavior of `BigDecimal.longValueExact()`, if decimal is equal to the value
+        // casted to decimal128 with scale 6, then no rounding is necessary.
+        val isEqual = withResource(input.getBase.castTo(decimal128Type)) { casted =>
+          withResource(casted.castTo(decimal128TypeAllScale)) { castedAllScale =>
+            withResource(input.getBase.castTo(decimal128TypeAllScale)) { original =>
+              original.equalTo(castedAllScale)
+            }
+          }
+        }
+        val roundUnnecessity = withResource(isEqual) { isEqual =>
+          withResource(isEqual.all()) { all =>
+            all.getBoolean()
+          }
+        }
+        if (!roundUnnecessity) {
+          throw new ArithmeticException("Rounding necessary")
+        }
+        val mul = withResource(input.getBase.castTo(decimal128Type)) { decimal128 =>
+          withResource(Scalar.fromLong(DateTimeConstants.MICROS_PER_SECOND)) { scalar =>
+            decimal128.mul(scalar, decimal128Type)
+          }
+        }
+        // Match the behavior of `BigDecimal.longValueExact()`:
+        val greaterThanMax = withResource(Scalar.fromLong(Long.MaxValue)) { longMax =>
+            mul.greaterThan(longMax)
+        }
+        val largerThanLongMax = withResource(greaterThanMax) { greaterThanMax =>
+          withResource(greaterThanMax.any()) { any =>
+            any.getBoolean()
+          }
+        }
+        lazy val smallerThanLongMin: Boolean = {
+          val lessThanMin = withResource(Scalar.fromLong(Long.MinValue)) { longMin =>
+              mul.lessThan(longMin)
+          }
+          withResource(lessThanMin) { lessThanMin =>
+            withResource(lessThanMin.any()) { any =>
+              any.getBoolean()
+            }
+          }
+        }
+        if (largerThanLongMax || smallerThanLongMin) {
+          mul.close()
+          throw new java.lang.ArithmeticException("Overflow")
+        }
+        withResource(mul) { mul =>
+          withResource(mul.castTo(DType.INT64)) { longs =>
+            longs.asTimestampMicroseconds()
+          }
+        }
+      }
+    case _ =>
```

Comment:
List the possible remaining types here. And use this `_` for unmatched types with throwing an exception saying "unsupported types $type"